net/portmapper/upnp.go

```diff
@@ -0,0 +1,208 @@
+// but the client field may be shared across mapping instances.
+type upnpMapping struct {
+	gw       netaddr.IP
+	external netaddr.IPPort
+	internal netaddr.IPPort
+	useUntil time.Time
+	// client is a connection to a upnp device, and may be reused across different UPnP mappings.
+	client upnpClient
+}
+
+func (u *upnpMapping) isCurrent() bool                { return u.useUntil.After(time.Now()) }
+func (u *upnpMapping) validUntil() time.Time          { return u.useUntil }
+func (u *upnpMapping) externalIPPort() netaddr.IPPort { return u.external }
+func (u *upnpMapping) release(ctx context.Context) {
+	u.client.DeletePortMapping(ctx, "", u.external.Port(), "udp")
+}
+
+// upnpClient is an interface over the multiple different clients exported by goupnp,
+// exposing the functions we need for portmapping. They are auto-generated from XML-specs.
+type upnpClient interface {
+	AddPortMapping(
+		ctx context.Context,
+		remoteHost string,
+		externalPort uint16,
+		protocol string,
+		internalPort uint16,
+		internalClient string,
+		enabled bool,
+		portMappingDescription string,
+		// Lease duration in seconds exposed by goupnp.
+		leaseDurationSec uint32,
+	) (err error)
+
+	DeletePortMapping(ctx context.Context, remoteHost string, externalPort uint16, protocol string) error
+	GetExternalIPAddress(ctx context.Context) (externalIPAddress string, err error)
+}
+
+const tsPortMappingDesc = "tailscale-portmap"
+
+// addAnyPortMapping abstracts over different UPnP client connections, calling the available
+// AddAnyPortMapping call if available for WAN IP connection v2, otherwise defaulting to the old
+// behavior of calling AddPortMapping with port = 0 to specify a wildcard port.
+func addAnyPortMapping(
+	ctx context.Context,
+	upnp upnpClient,
+	externalPort uint16,
+	internalPort uint16,
+	internalClient string,
+	leaseDuration time.Duration,
+) (newPort uint16, err error) {
+	if upnp, ok := upnp.(*internetgateway2.WANIPConnection2); ok {
+		return upnp.AddAnyPortMapping(
+			ctx,
+			"",
+			externalPort,
+			"udp",
+			internalPort,
+			internalClient,
+			true,
+			tsPortMappingDesc,
+			uint32(leaseDuration.Seconds()),
+		)
+	}
+	err = upnp.AddPortMapping(
+		ctx,
+		"",
+		externalPort,
+		"udp",
+		internalPort,
+		internalClient,
+		true,
+		tsPortMappingDesc,
+		uint32(leaseDuration.Seconds()),
+	)
+	return internalPort, err
+}
+
+// getUPnPClients gets a client for interfacing with UPnP, ignoring the underlying protocol for
+// now.
+// Adapted from https://github.com/huin/goupnp/blob/master/GUIDE.md.
+func getUPnPClient(ctx context.Context) (upnpClient, error) {
+	tasks, _ := errgroup.WithContext(ctx)
+	// Attempt to connect over the multiple available connection types concurrently,
+	// returning the fastest.
+	var ip1Clients []*internetgateway2.WANIPConnection1
+	tasks.Go(func() error {
+		var err error
+		ip1Clients, _, err = internetgateway2.NewWANIPConnection1Clients()
+		return err
```

Comment:
do you want the failure of any one of these three to cancel everything? As written, that's what it kinda looks like you're trying to do. (You're not using errgroup.Group in quite the normal way, though, so these returned errors aren't actually canceling anyting. But once you start using the ctx, then you'd want to return nil here if it could just be some sort of "nothing exists" error... I don't know what it returns)

Reply:
Yea, none of these `New*Clients` calls actually takes a context so it's a bit hard to handle properly. I was hoping to make it so that if all of them error, it returns any of the errors. If there is at least one success, just take that client and ignore other errors.